File: src/tools/pmd-cpd/eval-repos/synthetic/go/no_duplication.go
// Package synthetic contains Go files for CPD testing with no code duplication.
package synthetic

import (
	"math"
	"sort"
)

// CalculateFibonacci calculates the nth Fibonacci number.
func CalculateFibonacci(n int) int64 {
	if n <= 1 {
		return int64(n)
	}
	var a, b int64 = 0, 1
	for i := 2; i <= n; i++ {
		a, b = b, a+b
	}
	return b
}

// IsPrime checks if a number is prime.
func IsPrime(num int) bool {
	if num < 2 {
		return false
	}
	if num == 2 {
		return true
	}
	if num%2 == 0 {
		return false
	}
	for i := 3; i <= int(math.Sqrt(float64(num))); i += 2 {
		if num%i == 0 {
			return false
		}
	}
	return true
}

// BinarySearch performs binary search on a sorted array.
func BinarySearch(arr []int, target int) int {
	left, right := 0, len(arr)-1
	for left <= right {
		mid := (left + right) / 2
		if arr[mid] == target {
			return mid
		}
		if arr[mid] < target {
			left = mid + 1
		} else {
			right = mid - 1
		}
	}
	return -1
}

// DataProcessor processes numerical data.
type DataProcessor struct {
	data      []float64
	processed bool
}

// NewDataProcessor creates a new DataProcessor.
func NewDataProcessor(data []float64) *DataProcessor {
	return &DataProcessor{data: data}
}

// Normalize normalizes the data to 0-1 range.
func (dp *DataProcessor) Normalize() []float64 {
	if len(dp.data) == 0 {
		return []float64{}
	}
	min, max := dp.data[0], dp.data[0]
	for _, v := range dp.data {
		if v < min {
			min = v
		}
		if v > max {
			max = v
		}
	}
	if max == min {
		result := make([]float64, len(dp.data))
		for i := range result {
			result[i] = 0.5
		}
		return result
	}
	result := make([]float64, len(dp.data))
	for i, v := range dp.data {
		result[i] = (v - min) / (max - min)
	}
	return result
}

// Statistics contains computed statistics.
type Statistics struct {
	Mean   float64
	Median float64
	Std    float64
}

// ComputeStatistics computes basic statistics on the data.
func (dp *DataProcessor) ComputeStatistics() Statistics {
	if len(dp.data) == 0 {
		return Statistics{}
	}
	n := float64(len(dp.data))
	var sum float64
	for _, v := range dp.data {
		sum += v
	}
	mean := sum / n

	sorted := make([]float64, len(dp.data))
	copy(sorted, dp.data)
	sort.Float64s(sorted)
	var median float64
	if len(sorted)%2 != 0 {
		median = sorted[len(sorted)/2]
	} else {
		median = (sorted[len(sorted)/2-1] + sorted[len(sorted)/2]) / 2
	}

	var varianceSum float64
	for _, v := range dp.data {
		varianceSum += (v - mean) * (v - mean)
	}
	std := math.Sqrt(varianceSum / n)

	return Statistics{Mean: mean, Median: median, Std: std}
}

File: src/tools/pmd-cpd/eval-repos/synthetic/python/no_duplication.py
"""Python file with no code duplication - all functions are unique."""


def calculate_fibonacci(n: int) -> int:
    """Calculate the nth Fibonacci number using iteration."""
    if n <= 1:
        return n
    a, b = 0, 1
    for _ in range(2, n + 1):
        a, b = b, a + b
    return b


def is_prime(num: int) -> bool:
    """Check if a number is prime."""
    if num < 2:
        return False
    if num == 2:
        return True
    if num % 2 == 0:
        return False
    for i in range(3, int(num**0.5) + 1, 2):
        if num % i == 0:
            return False
    return True


def binary_search(arr: list[int], target: int) -> int:
    """Perform binary search on a sorted array."""
    left, right = 0, len(arr) - 1
    while left <= right:
        mid = (left + right) // 2
        if arr[mid] == target:
            return mid
        elif arr[mid] < target:
            left = mid + 1
        else:
            right = mid - 1
    return -1


def merge_sort(arr: list[int]) -> list[int]:
    """Sort an array using merge sort algorithm."""
    if len(arr) <= 1:
        return arr
    mid = len(arr) // 2
    left = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])
    return merge(left, right)


def merge(left: list[int], right: list[int]) -> list[int]:
    """Merge two sorted arrays into one sorted array."""
    result = []
    i = j = 0
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1
    result.extend(left[i:])
    result.extend(right[j:])
    return result


class DataProcessor:
    """A class for processing various types of data."""

    def __init__(self, data: list):
        self.data = data
        self.processed = False

    def normalize(self) -> list[float]:
        """Normalize numeric data to 0-1 range."""
        if not self.data:
            return []
        min_val = min(self.data)
        max_val = max(self.data)
        if max_val == min_val:
            return [0.5] * len(self.data)
        return [(x - min_val) / (max_val - min_val) for x in self.data]

    def compute_statistics(self) -> dict:
        """Compute basic statistics on the data."""
        if not self.data:
            return {"mean": 0, "median": 0, "std": 0}
        n = len(self.data)
        mean = sum(self.data) / n
        sorted_data = sorted(self.data)
        median = sorted_data[n // 2] if n % 2 else (sorted_data[n // 2 - 1] + sorted_data[n // 2]) / 2
        variance = sum((x - mean) ** 2 for x in self.data) / n
        return {"mean": mean, "median": median, "std": variance**0.5}
